Device structure

webdjoe opened the pull request:
Restructure device classes:

- Add `DeviceContainer` object that inherits from MutableSet to hold all devices. Includes helper functions to add and remove new devices from device list API response. This is held in the `VeSync.devices` attribute.
- All API request and responses are modeled with Dataclasses, using mashumaro/orjson for serialization and deserialization
- Separate out all state attributes into a separate object held in the `device.state` attribute. The state attribute stores all device information that can change
- Add constants for all modes, features, etc. `pyvesync.const`
- Organize devices by `product_type` attribute - purifiers, humidifiers, fans, outlets, switches, bulbs, etc.
- Add base device and base state class for each product type. These hold all attributes whether supported or not by all devices. The attributes that are not supported by all devices use the `IntFlag.NOT_SUPPORTED` or `StrFlag.NOT_SUPPORTED` depending on type.
- Standardize all methods and properties across classes and devices. Methods that only have on/off functionality start with `toggle_` and methods that can have a range of settings start with `set_`.

webdjoe committed:
Initial Refactor Commit

webdjoe committed:
Add device helper getattr to make testing easier

cdnninja commented on src/pyvesync/data_models/outlet_models.py:
```diff
@@ -0,0 +1,146 @@
+    traceId: str
+    userCountryCode: str
+    debugMode: bool
+    homeTimeZone: str
```
Should this be a time zone data type from datetime?

webdjoe replied:
This comes from API, so it should be in format API expects. No need to deserialize it into an object.

cdnninja commented on src/pyvesync/data_models/purifier_models.py:
```diff
@@ -0,0 +1,281 @@
+    AQLevel: int
+    PM25: int
+    screenState: int
+    childLockSwitch: int
```
Bool?

webdjoe replied:
All vital/everest on/off toggling is 1/0

cdnninja reviewed (changes requested):
I haven't gone through it all but a few things to start.  Great work on this!!

cdnninja commented:
Once this is merged I will give it a go with home assistant. I am sure I will learn more with that!

webdjoe committed:
Continued Refactoring

Changed module and class names to shorten references.

Added more data validation to API requests.

Fix device bugs

webdjoe committed:
Timers and optimizations

Add timer methods to devices and optimize for memory usage and speed.

webdjoe committed:
Documentation

webdjoe committed:
Autostopswitch

webdjoe committed:
Aura thermostat

webdjoe committed:
Update device_container.py
